head/head.go

```diff
@@ -137,19 +137,35 @@ func NewHead(ctx context.Context, options ...opts.Option) (*Head, chan Bootstrap
 	}
 
 	var cachingProviderStore *hproviders.CachingProviderStore
-	if cfg.ProvidersFinder != nil {
-		cachingProviderStore = hproviders.NewCachingProviderStore(providerStore, cfg.ProvidersFinder, nil)
-		providerStore = cachingProviderStore
+	if cfg.ProvidersFinder != nil && cfg.StoreTheIndexAddr == "" {
+		providerStore = hproviders.NewCachingProviderStore(providerStore, providerStore, cfg.ProvidersFinder, nil)
 	}
+	if cfg.ProvidersFinder != nil && cfg.StoreTheIndexAddr != "" {
```

Comment:
I think this sequence of if statements is correct, but for future reference: When your intention is to have mutually-exclusive cases, both for readability and safety, it is best to capture them either as a switch statement, or with "if else" chain. Here switch statement would be best:

switch {
case cfg.ProvidersFinder != nil && cfg.StoreTheIndexAddr == "":
  ...
  break
case cfg.ProvidersFinder != nil && cfg.StoreTheIndexAddr != "":
  ...
  break
case cfg.ProvidersFinder == nil && cfg.StoreTheIndexAddr != "":
  ...
  break
default:
  ... something is not right ...
}